### ISSUE TITLE
warn impl Deref

### ISSUE THREAD

**Issue by cathaysia** [rust-lang/rust-clippy#13665]
### What it does

[deref](https://doc.rust-lang.org/std/ops/trait.Deref.html) is usually implemented for transparent types such as pointers, but some people use it for other purposes, such as polymorphism, etc.

Without the help of lsp, it is difficult for the reader to know where this method comes from.

- [deref anti-pattern](https://rust-unofficial.github.io/patterns/anti_patterns/deref.html)

### Advantage

More clear code.

### Drawbacks

If possible, we should warn users when they implement deref

### Example

```rust
struct Parent {}

impl Parent {

   fn some_op(&self) {}
}

struct Child {
    parent: Parent 
}

impl Deref<Parent> for Child {
       // ^ warning: DON'T implement Deref
  // ...
}

```

Could be written as:

```rust
struct Parent {}

impl Parent {

   fn some_op(&self) {}
}

struct Child {
    parent: Parent 
}

impl Child {
    fn some_op(&self) {
       self.parent.some_op(&self)
   }
}
```
